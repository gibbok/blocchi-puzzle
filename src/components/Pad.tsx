import React from 'react';
import styled from 'styled-components';
import { wood } from '../assets/images';

const PadStyled = styled.div`
  position: relative;
  width: 8rem;
  height: 8rem;
  border-radius: 4rem;
  font-size: 5rem;
  font-family: sans-serif;
  box-shadow: inset 0.1rem 0.1rem 0.3rem 0 rgba(0, 0, 0, 0.4);
  transform: rotate(90deg);
  clip-path: circle(50% at 50% 50%);
  border: 0.2rem solid rgba(0, 0, 0, 0.1);
  box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.12);
`;

const PadBackground = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-image: url(${wood});
  clip-path: circle(50% at 50% 50%);
`;

const Shadow = styled.div`
  filter: drop-shadow(-0.2rem 0.3rem 0.3rem rgba(50, 50, 0, 0.4));
`;

type Props = Readonly<{
  icon: React.ReactNode;
  onClick: () => void;
}>;

export function Pad({ icon, onClick }: Props) {
  return (
    <Shadow>
      <PadStyled onClick={onClick}>
        <PadBackground>{icon}</PadBackground>
      </PadStyled>
    </Shadow>
  );
}
