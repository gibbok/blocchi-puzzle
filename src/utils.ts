import { Tetro, Board } from './types';

export const logger = (x: Tetro | Board) => {
  console.log(x.reduce((acc, v) => `${acc} ${v.concat()} \n`, ''));
};
