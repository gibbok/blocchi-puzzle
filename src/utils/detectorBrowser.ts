import { detect } from 'detect-browser';

const getPathRelease = (version: string) => version.split('.')[0];

export const detectorBrowser = (): boolean => {
  const browser = detect();
  if (browser && browser.name && browser.version) {
    const name = browser.name;
    const version = browser.version;
    const path = Number(getPathRelease(version));

    switch (name) {
      case 'chrome':
      case 'firefox':
        return path > 80;
      case 'safari':
        return path > 13;
      case 'opera':
        return path > 62;
      default:
        return false;
    }
  } else {
    return false;
  }
};
