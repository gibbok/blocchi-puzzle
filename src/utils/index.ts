export * from './data.support.test';
export * from './logger';
