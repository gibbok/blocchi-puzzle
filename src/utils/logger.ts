import { Tetro, Board } from '../game/types';

export const logger = (x: Tetro | Board): void => {
  console.log(x.reduce((acc, v) => `${acc} ${v.concat()} \n`, ''));
};
