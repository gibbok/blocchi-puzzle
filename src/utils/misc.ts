import { IO, io } from 'fp-ts/lib/IO';
import { randomInt } from 'fp-ts/lib/Random';

export const getRandomValueFromStringEnum = <T>(strEnum: T): IO<T[keyof T]> => {
  const rndInt = randomInt(0, Object.keys(strEnum).length - 1)();
  const rndValue = Object.keys(strEnum)[rndInt] as keyof T;
  return io.of(strEnum[rndValue]);
};
