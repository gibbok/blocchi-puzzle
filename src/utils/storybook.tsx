import React from 'react';
import GlobalStyle from '../components/GlobalStyle';
import styled from 'styled-components';
import { wood } from '../assets/images';

/**
 * Decorate a story with glocal styles.
 * @param story Story
 */
export const globalStylesDecorator = (story: () => unknown): JSX.Element => (
  <React.Fragment>
    <GlobalStyle />
    {story()}
  </React.Fragment>
);

const Bkg = styled.div`
  background-image: url(${wood});
  width: 100vw;
  height: 100vh;
`;

/**
 * Decorate a story with a page background.
 * @param story Story
 */
export const pageBackground = (story: () => React.ReactNode): JSX.Element => <Bkg>{story()}</Bkg>;
