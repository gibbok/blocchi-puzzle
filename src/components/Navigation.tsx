import React from 'react';
import styled from 'styled-components';
import { PadRotate, PadLeft, PadRight, PadDown } from './Pad';

const NavigationStyled = styled.div`
  padding: 1rem;
  display: flex;
  justify-content: space-between;
  width: 100%;
`;

type Props = Readonly<{
  onClickLeft: () => void;
  onClickRight: () => void;
  onClickDown: () => void;
  onClickRotate: () => void;
}>;

export function Navigation({
  onClickLeft,
  onClickRight,
  onClickDown,
  onClickRotate,
}: Props): JSX.Element {
  return (
    <NavigationStyled>
      <PadLeft onClick={onClickLeft} />
      <PadRotate onClick={onClickRotate} />
      <PadDown onClick={onClickDown} />
      <PadRight onClick={onClickRight} />
    </NavigationStyled>
  );
}
