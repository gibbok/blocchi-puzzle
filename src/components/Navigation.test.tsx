import React from 'react';
import { Navigation } from './Navigation';
import renderer from 'react-test-renderer';
import { noop } from '../utils';
import { PadLeft, PadRight, PadDown, PadRotate } from './Pad';

describe('<Navigation />', () => {
  const left = jest.fn();
  const right = jest.fn();
  const down = jest.fn();
  const rotate = jest.fn();

  const tree = renderer.create(
    <Navigation onClickLeft={left} onClickRight={right} onClickDown={down} onClickRotate={rotate} />
  );
  const instance = tree.root;

  it('should render Navigation', () => {
    tree.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('it should execute left callback on click', () => {
    const pad = instance?.findByType(PadLeft);
    pad.props.onClick();
    expect(left).toBeCalled();
  });

  it('it should execute right callback on click', () => {
    const pad = instance?.findByType(PadRight);
    pad.props.onClick();
    expect(right).toBeCalled();
  });

  it('it should execute down callback on click', () => {
    const pad = instance?.findByType(PadDown);
    pad.props.onClick();
    expect(down).toBeCalled();
  });

  it('it should execute rotate callback on click', () => {
    const pad = instance?.findByType(PadRotate);
    pad.props.onClick();
    expect(rotate).toBeCalled();
  });
});
