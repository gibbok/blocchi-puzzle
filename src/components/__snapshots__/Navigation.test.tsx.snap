// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should render Navigation 1`] = `
<div
  className="sc-AxgMl MMksN"
>
  <div
    className="sc-AxhUy ehyRyD"
  >
    <div
      className="sc-AxirZ lcEoVe"
      onClick={[MockFunction]}
    >
      <div
        className="sc-AxiKw cWlzsq"
      >
        <div
          className="sc-AxhCb kiYDUb"
        >
          <svg
            aria-hidden="true"
            className="sc-AxjAm gXIMwt"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
            />
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-AxhUy ehyRyD"
  >
    <div
      className="sc-AxirZ lcEoVe"
      onClick={[MockFunction]}
    >
      <div
        className="sc-AxiKw cWlzsq"
      >
        <div
          className="sc-AxhCb kiYDUb"
        >
          <svg
            aria-hidden="true"
            className="sc-AxjAm gXIMwt"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M15.55 5.55L11 1v3.07C7.06 4.56 4 7.92 4 12s3.05 7.44 7 7.93v-2.02c-2.84-.48-5-2.94-5-5.91s2.16-5.43 5-5.91V10l4.55-4.45zM19.93 11a7.906 7.906 0 00-1.62-3.89l-1.42 1.42c.54.75.88 1.6 1.02 2.47h2.02zM13 17.9v2.02c1.39-.17 2.74-.71 3.9-1.61l-1.44-1.44c-.75.54-1.59.89-2.46 1.03zm3.89-2.42l1.42 1.41c.9-1.16 1.45-2.5 1.62-3.89h-2.02c-.14.87-.48 1.72-1.02 2.48z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-AxhUy ehyRyD"
  >
    <div
      className="sc-AxirZ lcEoVe"
      onClick={[MockFunction]}
    >
      <div
        className="sc-AxiKw cWlzsq"
      >
        <div
          className="sc-AxhCb kiYDUb"
        >
          <svg
            aria-hidden="true"
            className="sc-AxjAm gXIMwt"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-AxhUy ehyRyD"
  >
    <div
      className="sc-AxirZ lcEoVe"
      onClick={[MockFunction]}
    >
      <div
        className="sc-AxiKw cWlzsq"
      >
        <div
          className="sc-AxhCb kiYDUb"
        >
          <svg
            aria-hidden="true"
            className="sc-AxjAm gXIMwt"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
            />
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
